### ISSUE TITLE
Joints of SimulationBody "break" (unexpected behaviour)

### ISSUE THREAD

**Issue by DubWa91** [dyn4j/dyn4j#126]
Hi William,

I have currently a quite strange behaviour in my physics simulation, because it seems like the `Joint`s I set "breake" during the simulation.

Please see the following video (GIF):

![Broken_joints](https://user-images.githubusercontent.com/34217758/91044354-70d60f80-e615-11ea-9124-17ac2038ca39.gif)

---

Here a quick summary what I am doing:

1) I have this rough plan of my simulation

```
/** Setup the layout simulation */
LayoutPlanningLauncher layoutSimulation = new LayoutPlanningLauncher(this.contentManager, pixelPerMeterScale, frameSize);

// Start the layout simulation for visualization purpose
layoutSimulation.startSimulation()

// Setup the static layout objects (border, input/output, interferingcontour)
layoutSimulation.initializeLayout();

// Get the generated solutions
ArrayList<AssemblyLine> assemblyLineSolutions = this.contentManager.getSolutionManager().getAssemblyLineSolutions();

// Loop over the generated solutions
for (AssemblyLine assemblyLine : assemblyLineSolutions)
{
	// Model the workstations / resources
	layoutSimulation.modelObjects(assemblyLine);
	
	// Place the workstations / resources
	layoutSimulation.arrangeWorkstations(assemblyLine);
	
	// Loop for equilibrium
	layoutSimulation.loop();
	
	// Clear the world before proceed with the next solution
	layoutSimulation.clearWorld();
}
```

2) In the method `startSimulation()`, I make the SimulationFrame visible and start the demon thread (like in your examples). Within the demon thread I have the `world.update()` method, as well as the rendering of `SimulationBody`

3) In the method `initializeLayout()`, I set static `SimulationBody` with `MassType.INFINITE` (e.g. the `Links` to define the black rectangle in the GIF)

4) In the method `modelObjects()`, I model `SimulationBody`s, which are placed within the method `arrangeWorkstations()`.

5) In the method `arrangeWorkstations()`, I place `SimulationBody`s within the layout (see 3) and apply different `Joint`s to them (e.g. `RopeJoint` -> see blue lines in the GIF). Additionally I add some `Listener` to monitor or observe some conditions.

6) In the method 'loop()' I observe an overall "manager" which obeserves a specific number of defined `Listener` (so called equilibrium)

```
System.out.println("Start of 'Endless Loop' to find system equilibrium");
		
// Loop, till the world updating is stopped through the ListenersManager
while(!this.paused)
{
	// Check if the listener nofify about equilibrium
	boolean isEquilibrium = this.listenerManager.isEquilibrium();
	
	if(isEquilibrium == true)
	{
		System.out.println("Equilibrium found -> Stop simulation");
		this.paused = true;
	}
	else
	{
		// Do nothing -> normal SimulationLoop
	}
}

System.out.println("Start of methods after Equilibrium -> still rendering.");

// TODO: Adjust methods
while(true)
{
	// Endless loop for testing
}
```

---

That's basically a rough explanaition, what I am doing.

Based on that description, can you give me a hint, what is going wrong so that the `Joint`s are breaking and the collsion detection of `SimulationBody`s doesn't work anymore (in the video the `SimulationBody`s go all through the constraining border -> `Link` set)?

If necessary I can give of course more information about the internal process of the above described methods :)

---

Edit: It doesn't show any errors in the console

And the behaviour depends on the simulation run. Sometimes (hard to catch via video) the `Joint`s still exist and jump wild in the simulationframe, without considering any `SimulationBody` or `BodyFixtures`

**Comment by wnbittle** [comment]
Is there a way you can reproduce outside of your app?  Basically output the initial state of the bodies, shapes, joints, etc so we can isolate the problem.

Is there a simpler scenario where this happens (less bodies and joints)?

I see that some bodies, when it explodes, are sent outside the bounds, and once there the joints are effectively disabled.  That explains the smooth running after it explodes.

It wasn't clear from the code, but it seemed like you were modifying the state of the world on a different thread then where you render.  There's lots of pitfalls with that.  Can you confirm?

**Comment by DubWa91** [comment]
> Is there a way you can reproduce outside of your app? Basically output the initial state of the bodies, shapes, joints, etc so we can isolate the problem.

I made a smaller example with the same code logic (method calling, structure etc.) and there everything went fine.
But I will use my existing code and try make a different project out of it so that you can test it also. But this will take some time (I think I can provide the code tomorrow)

---

> Is there a simpler scenario where this happens (less bodies and joints)?

Here is my simplier example with the same code logic

![Spring_approximation](https://user-images.githubusercontent.com/34217758/91269801-fb368480-e777-11ea-9fc5-3966af8386b0.gif)

To explain my simplier scenario a bit:
1) I generate a few bodies and connect them with `RopeJoint`s
2) I connect 2 `SimulationBody` with a `DistanceJoint` (spring behaviour)
3) To reduce the impulses, which are applied to the `SimulationBody`s through the instant set `DistanceJoint`, I add a `StepAdapter`, which changes the defined rest length (distance) of the `DistanceJoint` every x simulation steps
4) After the target rest length is completly set to the `DistanceJoint` I remove the `StepAdapter` from the world

My goal is to use that `DistanceJoint` to change the positions of the connected `SimulationBody` till specific criteria are met. I hope you can follow my explanation :-)

---

> I see that some bodies, when it explodes, are sent outside the bounds, and once there the joints are effectively disabled. That explains the smooth running after it explodes.

I guess that is, because I defined an `OutOfBoundsListener` which has the dimensions of the visiable black border (set of `Link`s). But I don't understandy why to collision of the `SimulationBody` inside the layout is disabled.

---

> It wasn't clear from the code, but it seemed like you were modifying the state of the world on a different thread then where you render. There's lots of pitfalls with that. Can you confirm?

Basically the updating of the world and the rendering is in the same thread (like [here ](https://github.com/dyn4j/dyn4j-samples/blob/master/src/main/java/org/dyn4j/samples/framework/SimulationFrame.java).

BUT: I start the demon thread (with the continous world updating and rendering) in `layoutSimulation.startSimulation()`. And then I have my loops, where I add new `SimulationBody`s, `Joint`s and `Listener`, because I don't have the possibility to define everything in advance and couldn't figuere out a better code structure... (as you can see above in my smaller example it worked)

**Comment by DubWa91** [comment]
Hi William,

I am very sorry about the late reply / update, but it was quite hard for me to reduce my existing project to the most essential parts, where I use your physics library.

Please find attached the following file:
- [layouting.zip](https://github.com/dyn4j/dyn4j/files/5156329/layouting.zip) - my project

I wasn't able to upload my used JAVA libraries, but here is the link for the downloads:
- [JGraphT_1.4.0](https://jgrapht.org/)
- [GSON 2.8.5](https://github.com/google/gson)
- and of course your DNY4J library as jar 3.4.0

---

I wasn't able to export my created `SimulationBody` as JSON objects (some problems with abstract classes in your library). Therefore you will find some additional methods, which you can skip.
These are all methods, which model the `SimulationBody` objects and identify some relations between them, which will be translated in a later stage into `Joint`s.
Basically you can skip all methods inside `LayoutingLauncher -> layoutSimulation.modelObjects(assemblyLine);`

---

I think the problem is caused by the method `LayoutingLauncher -> layoutSimulation.arrangeWorkstations(assemblyLine);`.

Here I place my different `SimulationBody`s in the defined layout (`LayoutingLauncher -> layoutSimulation.initializeLayout();` and model and apply the previously identified `Joint`s.

From the logic, I loop different objects (`Workstation`), which contain different `SimulationBody`s (`Resource`) -> these have relations to each other (`Joint`s). At this level everything works.

Now, when I want to place the next `Workstation` and apply some `Joint`s between these "higher level" objects (Workstation-Workstation) I will get in a later stage of the simulation the above mentioned problems with the breaking `Joint`s.

=> `LayoutPlanningLauncher -> arrangeWorkstations(AssemblyLine assemblyLine) -> this.modelWorkstationRelations(assemblyLine, previousWorkstation, currentWorkstation);` [LayoutPlanningLauncher line 697].

And then the method for modelling:

`relationModeller.modelRelation(relation, simulationBodyA, simulationBodyB); ` => model/apply `DistanceJoint` (spring)

Extract of class `RelationModeller` with its method `modelTimeDistanceRelation((TimeDistanceRelation) relation, simulationBodyA, simulationBodyB);`

Starting from [RelationModeller line: 162]

```
		/*
		 * Create the Joint and add it to the world
		 */
		SpringJoint springJoint = new SpringJoint(simulationBodyA, simulationBodyB, globalAnchorA, globalAnchorB);
		springJoint.setCollisionAllowed(true);
		double initialRestLength = relation.getRequiredDistance();	// Based on inital setting
		springJoint.setDistance(initialRestLength);
		double initalDampingRatio = 0.1;	// Based on test runs
		springJoint.setDampingRatio(initalDampingRatio);
		double initalFrequency = 8.0;	// Based on test runs
		springJoint.setFrequency(initalFrequency);
		springJoint.setRenderJoint(true);	// Rendering
		springJoint.setJointColor(Color.GREEN);	// Rendering
		this.world.addJoint(springJoint);
		
		
		/*
		 * Since springs cause a lot of impule to SimulationBodys when set immediately we have to approximate the target distance
		 * with small steps each Simulation loop
		 */
		double approximationStepFrequency = 0.005;	// 50 mm
		SpringForceRelaxation springForceRelaxation = new SpringForceRelaxation(springJoint, approximationStepFrequency);
		springForceRelaxation.setSimulationStepNumber(20);
		this.world.addListener(springForceRelaxation);
		
		// Loop, till target distance is reached
		this.relaxationLoop(springForceRelaxation);
		
		
		/*
		 * Add a listener to check if the length is maintained
		 */
		JointLengthListener jointLengthListener = new JointLengthListener(springJoint);
		this.world.addListener(jointLengthListener);
		springJoint.addRelatedListener(jointLengthListener);	// Add listener to joint
```

Waiting for target distance reached:

```
/**
	 * This loop is active, as long as the target distance of a {@link SpringJoint} is not reached. <br>
	 * The approximation of the distance is done by the {@link Listener} {@link SpringForceRelaxation}.
	 *
	 * @param springForceRelaxation - The listener of class {@link SpringForceRelaxation}
	 */
	private void relaxationLoop(SpringForceRelaxation springForceRelaxation)
	{
		while (springForceRelaxation.isActive() == true)
		{
			// "Endless loop", till condition is achieved
		}
		
		this.world.removeListener(springForceRelaxation);
	}
```

Basically my goal is minimize the required space of all objects in the layout. Therefore I want to pull the different `Workstation`s and their `SimulationBody` as close togeather as possible but under some constraints (e.g. no overlapping or maintain specific distances -> maybe this will explain ALL my other questions so far :-D )  

---

I know it is very hard to grasp what I am doing here, but I coulnd't export or provide less information. I would really appreciate your help, because I am now struggeling with that problem a few weeks and have no idea how to fix it.

 Thank you very much in advance. Every small hint helps :-)

Best regards, 

Carsten

---

PS: I have to arrange `Workstation` by `Workstation` since there are some contraints from the previous one, which I have to consider for the following one. (just in case you wonder my loops)

**Comment by DubWa91** [comment]
Update from 2020-09-07:

I removed the method `relaxationLoop(SpringForceRelaxation springForceRelaxation)` of class `RelationModeller` and set only the `SpringJoint`s.

The initial distance value (= rest length) is set to the current distance between the 2 connected `SimulationBody`. So therfore no additional forces should be applied.

After a certain time my rendering collapses. That means some `SimulationBody` disappear from the view and I get console messages, that `SimulationBody`s xyz are outside the bounds (activated through my `LayoutBoundsListener`).

Hope this additional information helps.

**Comment by wnbittle** [comment]
Yeah, I apologize, but there's a lot in this project and I don't have the time to go through it all.  The best way for me to help is to serialize the scene in a way that I can reproduce the problem in the examples framework.  My goal is to determine if this is a dyn4j issue, a scene layout problem, etc.

The only recommendation I found is by moving the `layoutSimulation.startSimulation();` call to right before the `layoutSimulation.loop();` call in the `startLayoutEngine` method of the `LayoutingLauncher` class?  The goal with this change is to  eliminate the multi-threading issue - basically build the whole scene, then start rendering/checking (only) the layout for equilibrium.  Can you try this?

**Comment by DubWa91** [comment]
> Yeah, I apologize, but there's a lot in this project and I don't have the time to go through it all. The best way for me to help is to serialize the scene in a way that I can reproduce the problem in the examples framework. My goal is to determine if this is a dyn4j issue, a scene layout problem, etc.

I totally understand that you can not go through my whole code. I also didn't expect that. :-)

When you are talking about `examples framework`, you mean that one you provided at [GitHub](https://github.com/dyn4j/dyn4j-samples/blob/master/src/main/java/org/dyn4j/samples/framework/SimulationFrame.java)?

Would it help, when I try to polish the code the following way:

1) I place the `SimulationBody`s in their initial position (before all joints are set)
2) I provide a list (coded-list) with the order of `Joint`s which are applied

---

> The only recommendation I found is by moving the layoutSimulation.startSimulation(); call to right before the layoutSimulation.loop(); call in the startLayoutEngine method of the LayoutingLauncher class?

Thank you for that hint I will try it :-)

But I assume, that this will get me some problems with my intended workflow. Because that way, it is assumed, that I know in advance the initial position of `SimulationBody`s. Which basically isn't the case since I have to place them step-by-step with time-based optimization (you remember about the goal of reducing the distance between them).

But as I mentioned I will try it and inform you about the results.

**Comment by DubWa91** [comment]
Hi William,

I could successfully re-create the explained behaviour in a much smaller testing environment.

You can find my project here: [JointBreakExample.zip](https://github.com/dyn4j/dyn4j/files/5194575/JointBreakExample.zip)

I used the dyn4j version 3.4.0

---

**Initial placement of `SimulationBody`**

![grafik](https://user-images.githubusercontent.com/34217758/92588274-08d02c00-f299-11ea-8a9a-76cc923e9ee4.png)

---

**Joints:**

*  `PrismaticJoint`: Anchor & A
*  `RopeJoint`: A & B
*  `RopeJoint`: A & C
*  `PrismaticJoint`: A & D 
*  `RopeJoint`: D & E
*  `RopeJoint`: D & F

---

**Bug behaviour**

*  I apply all `Joint`s: Everything breaks and only 2 `SimulationBody` are within the layout
*  I remove only the `PrismaticJoint` A & D: Everything works
*  I remove only the `PrismaticJoint` Anchor & A : Everything works
*  I remove `SimulationBody` D, E, F and their `Joints`: Everything works
*  I remove `SimulationBody` D, E, F and their `Joints` and also the `PrismaticJoint` Anchor & A: `SimulationBody` A disappears (transform dz = NaN)

---

To monitor the different problems I created 2 listener

*  `TransformListener`: Prints **x** and **y** coordinate and **dz** rotation
*  `LayoutBoundsListener`: Print in console, when a `SimulationBody` is outside the bounds

---

I set on purpose the `RopeJoint`s to the upper lengths to match the existing problem from my main project.

You can start the mini application in `BugLauncher.java`

**Comment by wnbittle** [comment]
@DubWa91 Thanks for this - it highlighted the problem perfectly.  This is a bug introduced in 3.4.0 in the Transform class.  When the scene explodes, you can see the current transforms for the bodies missing are have all NaNs.  This was fixed in 4.0.0 in the massive #127 PR.

Can you update to 4.0.0 and try the more complex scenarios to confirm?

**Comment by DubWa91** [comment]
Hi William,

it took me some time, but now I can confirm, that for my "small" example, the new DYN4J library is working :-)

Thank you very much for that new version (at a first glance very confusing for me and a lot of class/method updating necessary but maybe that's because of my beginner level :-) ).

---

For my "big" example I am now struggeling with some errors about `java.util.ConcurrentModificationException` but this is (I think) caused, because I am adding some `SimulationBody` or `Joint`s during one code execution step (simulation step).

Basically because I have some execution methods "beside" the update thread (but this is definitely not a problem of the engine :-) ) But this isn't your problem ;)

---

Again, thank you very much for your patience with my questions
